classes/privacy/provider.php

```diff
@@ -0,0 +1,266 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * kuraCloud privacy provider.
+ *
+ * Implement support for describing and managing private data.
+ *
+ * @package    block_kuracloud
+ * @category   privacy
+ * @copyright  2020 KuraCloud
+ * @author     John Gee
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace block_kuracloud\privacy;
+
+defined('MOODLE_INTERNAL') || die();
+
+use \core_privacy\local\metadata\collection;
+use \core_privacy\local\request\contextlist;
+use \core_privacy\local\request\writer;
+use \core_privacy\local\request\userlist;
+use \core_privacy\local\request\approved_contextlist;
+use \core_privacy\local\request\approved_userlist;
+
+
+/**
+ * Delete user data for target user which is no longer associated with a valid course.
+ *
+ * @param  int $userid The target used.
+ */
+function kuracloud_delete_stale_data_for_user(int $userid) {
+    global $DB;
+
+    // Get list of stale records, which are no longer associated with a course.
+    $sql = "SELECT kc_users.id
+              FROM {block_kuracloud_users} kc_users
+         LEFT JOIN {block_kuracloud_courses} kc_courses on kc_users.remote_instanceid = kc_courses.remote_instanceid AND kc_users.remote_courseid = kc_courses.remote_courseid
+         LEFT JOIN {course} c on c.id = kc_courses.courseid
+             WHERE c.id IS NULL AND kc_users.userid = :userid
+    ";
```

Comment:
What happens when a user is part of a dead course and part of an alive course?

Reply:
There is a row per user per course, so this only finds the rows where the course is gone burgers.